### PR TITLE
Text generation

### PR DESCRIPTION
**Pull request by oninbo**

### CODE REVIEW

File: testing.py [hunk 1]
```diff
@@ -36,10 +38,61 @@ def test_reply_agent(agent_function: "agent's function to process input message"
     json_manager.write(test_output, test_output_file_name)
 
 
-TEST_NUMBERS = [1, 0, 2]
+def agent_testing() -> None:
```
Comment (UnoYakshi) [inline testing.py]: Please, consider using `unittests` (at least) module for such stuff.

File: testing.py [hunk 2]
```diff
@@ -36,10 +38,61 @@ def test_reply_agent(agent_function: "agent's function to process input message"
...
+    """Tests agents"""
```
Comment (UnoYakshi) [inline testing.py]: It's a duplicate of the method's name, not a document-like description.

File: testing.py [hunk 3]
```diff
@@ -36,10 +38,61 @@ def test_reply_agent(agent_function: "agent's function to process input message"
...
+                             os.path.join('data', 'tests', f'test{str(test_n)}.txt'),
+                             test_output_file_name=f'test_output_CONVERSATION_CONTROLLER_n_{str(test_n)}.txt')
+
+
+def text_generating(test_n: int = 0, all_replies: bool = False) -> None:
+    """
+    Tests text generation using text tests
+    :param test_n: # of the test
+    :param all_replies: indicates that all replies should be printed
+    :return: None
```
Comment (UnoYakshi) [inline testing.py]: It returns nothing, not None. Though `-> None` in annotation covers this case as well.

File: testing.py [hunk 4]
```diff
@@ -36,10 +38,61 @@ def test_reply_agent(agent_function: "agent's function to process input message"
...
+            if all_replies:
+                print('out:\t', replies)
+            else:
+                print('out:\t', random.choice(list(replies)))
+        else:
+            print('out:\t')
+        print()
+
+
+def dialog_imitation() -> None:
```
Comment (UnoYakshi) [inline testing.py]: I suppose it's a dialogue, not a dialog.

File: testing.py [hunk 5]
```diff
@@ -36,10 +38,61 @@ def test_reply_agent(agent_function: "agent's function to process input message"
...
+    """
+    Imitates dialog using text generation through console input
+    :return: None
+    """
+    base_path = os.path.join('data', 'text', 'speech_base.txt')
+    with open(base_path, 'r') as file:
+        text_gen = PartsOfSpeechTextGenerator(file.read())
+
+    while True:
```
Comment (UnoYakshi) [inline testing.py]: Please, confirm it should always run.

File: text_generation.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
+"""
+Module for text generating
```
Comment (UnoYakshi) [inline text_generation.py]: Please, provide more complete module docstring.

File: text_generation.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+
+LETTERS_REGEX = r'[^\W\d_]'
+END_PUNCT_REGEX = '[!.?]'
+
+LOGGER = logger.get_logger(__file__)
+
+
+def measure_run_time(func: Callable) -> Callable:
+    """
+    Wrapper for measuring run time of a given function
```
Comment (UnoYakshi) [inline text_generation.py]: In the case of Python, I think it's a decorator, not a wrapper. However, I don't see a big difference. Just FYI.

File: text_generation.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+        start_time = process_time()
+        result = func(*args, **kwargs)
+        print(func.__name__ + ' ' + str(process_time() - start_time))
+        return result
+    return call_and_measure
+
+
+class TextGenerator:
+    """
+    For text generation
```
Comment (UnoYakshi) [inline text_generation.py]: Provide more complete class docstring.

File: text_generation.py [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+    """
+    def __init__(self, base_text: str):
+        """
+        :param base_text: text which is used for generation
+        """
+
+        lines = base_text.split('\n')
+        sentences = []
+        for line in lines:
+            sentences += sent_tokenize(line)
```
Comment (UnoYakshi) [inline text_generation.py]: How about one-line solution?

File: text_generation.py [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+            result = result.replace(up, up[:3])
+
+        # remove paired symbols without pair
+        pair_puncts = {
+            "''": '``',
+            '(': ')',
+            '[': ']',
+            '{': '}',
+            '«': '»'
+        }
```
Comment (UnoYakshi) [inline text_generation.py]: I think it should be extracted as a constant somewhere.

File: text_generation.py [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+        # regex to check if the text is complete
+        check_regex = f'^{LETTERS_REGEX}.*{END_PUNCT_REGEX}$'
+        return re.search(check_regex, text)
+
+    def generate(self, start_words: List[str] = None,
+                 begin_links: Dict[str, Set[str]] = None,
+                 end_links: Dict[str, Set[str]] = None,
+                 max_word_num: int = None) -> Set[str]:
+        """
+        Generates texts
```
Comment (UnoYakshi) [inline text_generation.py]: Provide more complete docstring.

File: text_generation.py [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+                next_i = i + 1
+                if current_i < len(pronouns_and_verbs) - 1 \
+                        and pronouns_and_verbs[current_i][1] == PARTS_OF_SPEECH['personal pronoun'] \
+                        and pronouns_and_verbs[next_i][1] == PARTS_OF_SPEECH['verb']:
+                    pronoun = pronouns_and_verbs[current_i][0].lower()
+                    verb = pronouns_and_verbs[next_i][0].lower()
+                    func(pronoun, verb)
+
+    @staticmethod
+    def _get_word_ending(word: str) -> str:
```
Comment (UnoYakshi) [inline text_generation.py]: Why static method is protected?

File: text_generation.py [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+        :return: None
+        """
+        links = self._pronoun_verb_ending_links
+
+        def add_pronoun_verb_link(pronoun: str, verb: str) -> None:
+            """
+            Adds the link
+            :param pronoun:
+            :param verb:
+            :return: None
```
Comment (UnoYakshi) [inline text_generation.py]: Complete the docstring: parameters are unfilled.

File: text_generation.py [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+        :return: indication of validity of verbs
+        """
+        result = [True]
+
+        def update_check_result(pronoun: str, verb: str) -> None:
+            """
+            updates result value depending on verb ending check
+            :param pronoun:
+            :param verb:
+            :return: None
```
Comment (UnoYakshi) [inline text_generation.py]: Params are incomplete.

File: testing.py (hunk 3 above)
Comment (oninbo) [reply testing.py]: If it is not defined what the method returns, the method returns None by default in Python.

For example,
```
>>> def method(): pass
... 
>>> a = method()
>>> a is None
True
```

File: testing.py (hunk 5 above)
Comment (oninbo) [reply testing.py]: I didn't get it.

File: testing.py (hunk 3 above)
Comment (oninbo) [reply testing.py]: I think we discussed it earlier.

File: testing.py (hunk 4 above)
Comment (oninbo) [reply testing.py]: My dictionary is fine with that.

proof

File: text_generation.py [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+
+        # remove empty parenthesis
+        empty_pairs = re.findall('(' + re.escape('|'.join(self._PAIR_PUNCTS.keys())) + ')' + ' '
+                                 + '(' + re.escape('|'.join(self._PAIR_PUNCTS.values())) + ')', result)
+        for e_p in empty_pairs:
+            result = result.replace(e_p, ' ')
+
+        # make the letter before ending punctuation upper case
+        low_letters = re.findall(END_PUNCT_REGEX + ' ' + LETTERS_REGEX, result)
+        for l_l in low_letters:
```
[the remark this answers is not in the document]
Comment (oninbo) [reply text_generation.py]: Yes, it stands for `low_letter`. Should I change it?

File: text_generation.py [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+        """
+        Adds ending punctuation to the text
+        :param text: initial text without ending punctuation
+        :return: text with ending punctuation on its end
+        """
+        return text + random.choice(self._ends)
+
+    def _polish_text(self, text: str) -> str:
+        """
+        Removes unwanted symbols from the generated text
```
[the remark this answers is not in the document]
Comment (oninbo) [reply text_generation.py]: They are in the comments inside the function. But I can add a few examples to the docstring.

File: testing_reply_agents.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    Class for testing replying agents
+    """
+    def test_reply_agents(self) -> None:
+        """
+        Tests agents that can give replies on text messages
+        :return: None
+        """
+
+        for test_n in TEST_NUMBERS:
+            for agent_function in AGENTS_REPLY_FUNCTIONS:
```
[the remark this answers is not in the document]
Comment (oninbo) [reply testing_reply_agents.py]: Not sure if it is useful if I don't use `self.assert*` method. If the error happens the testing will crash anyway.

[the remark this answers is not in the document]
Comment (UnoYakshi) [reply testing_reply_agents.py]: Though, it shouldn't be stopped on the first crash encounter. And `self.subTest(..)` creates a separate context for each run.

File: text_generation.py [hunk 18]
```diff
@@ -181,11 +188,11 @@ def _polish_text(self, text: str) -> str:
         return self._add_end_punct(result) \
             if re.search('[^'+re.escape(string.punctuation)+']$', result) else result
 
-    def _evaluate_text(self, text: str) -> bool:
+    def _is_text_correct(self, text: str) -> bool:
         """
-        Checks if the text is complete
+        Checks if the text is correct
```
Comment (UnoYakshi) [inline text_generation.py]: It's better to define what is complete/correct.

**Pull request merged**